spec/controllers/catalog_controller_spec.rb

```diff
@@ -1291,60 +1291,28 @@
   end
 
   context "hiding columns" do
-    let(:user) { FactoryBot.create(:user) }
-
-    before do
-      Tenant.seed
-      EvmSpecHelper.local_miq_server
-    end
-
-    let!(:service_template) { FactoryBot.create(:service_template, :description => 'XXX', :tenant => Tenant.root_tenant) }
-
-    let!(:report) do
-      FactoryGirl.create(:miq_report,
-                         :name        => 'Catalog Items',
-                         :db          => 'ServiceTemplate',
-                         :title       => 'Catalog Items',
-                         :cols        => %w[name description type_display],
-                         :col_order   => %w[name description tenant.name type_display],
-                         :headers     => %w[Name Description Tenant Type],
-                         :col_options => {"tenant.name" => {:display_method => :user_super_admin?}})
-    end
-
-    it "renders show_list and does not include hidden column(hidden by display method)" do
-      allow(controller).to receive(:render)
-
-      login_as user
-
-      expect(controller).to receive(:get_db_view).and_return(report)
-      controller.send(:report_data)
-      view_hash = controller.send(:view_to_hash, assigns(:view))
-
-      headers = view_hash[:head].map { |x| x[:text] }.compact
-      expect(headers).to match_array(%w[Name Description Type])
-      expect(headers).not_to include('Tenant')
-
-      first_row_values = view_hash[:rows][0][:cells].map { |x| x[:text] }.compact
-      expect(first_row_values).to match_array([service_template.name, service_template.description, service_template.type_display])
-      expect(first_row_values).not_to include(service_template.tenant.name)
-    end
-
-    let(:user_admin) { FactoryBot.create(:user_admin) }
-
-    it "renders show_list and includes all columns" do
-      allow(controller).to receive(:render)
-
-      login_as user_admin
-
-      expect(controller).to receive(:get_db_view).and_return(report)
-      controller.send(:report_data)
-      view_hash = controller.send(:view_to_hash, assigns(:view))
-
-      headers = view_hash[:head].map { |x| x[:text] }.compact
-      expect(headers).to match_array(%w[Name Description Tenant Type])
-
-      first_row_values = view_hash[:rows][0][:cells].map { |x| x[:text] }.compact
-      expect(first_row_values).to match_array([service_template.name, service_template.description, service_template.tenant.name, service_template.type_display])
+    it_behaves_like '#hidingColumns'do
+      before do
+        Tenant.seed
+        EvmSpecHelper.local_miq_server
+      end
+
+      let!(:service_template) { FactoryBot.create(:service_template, :description => 'XXX', :tenant => Tenant.root_tenant) }
```

Comment:
I think the `!` is unnecessary here, please look up the difference between `let` and `let!` and minimize the use of `!` based on the assessment.

Reply:
http://www.betterspecs.org/#let

Reply:
It doesn't work without the `!`, so I guess it is necessary. Although I'm not sure that the `!` is necessary in the second appearance ` let!(:report) do ...` (line 1305)

Reply:
Well, this sounds like:
![jJwgH4k](https://user-images.githubusercontent.com/649130/61722548-c54ff100-ad6a-11e9-9007-e748164e6dbc.jpg)

I'll show you how to deal with `let` and `let!` tomorrow and how can you transitively clean this up.